dll/win32/shell32/wine/changenotify.c

```diff
@@ -165,30 +153,26 @@ static void DeleteNode(LPNOTIFICATIONLIST item)
 
     /* free the item */
     for (i=0; i<item->cidl; i++)
-#ifdef __REACTOS__
-    {
-        QueueUserAPC(_RequestTermination, m_hThread, (ULONG_PTR) &item->apidl[i] );
-        WaitForSingleObjectEx(m_hThread, 100, FALSE);
-#endif
         SHFree((LPITEMIDLIST)item->apidl[i].pidl);
-#ifdef __REACTOS__
-        SHFree(item->apidl[i].buffer);
-        SHFree(item->apidl[i].backBuffer);
-    }
-#endif
     SHFree(item->apidl);
     SHFree(item);
 }
+#endif  /* ndef __REACTOS__ */
 
 void InitChangeNotifications(void)
 {
-#ifdef __REACTOS__
-    m_hThread = NULL;
-#endif
 }
 
 void FreeChangeNotifications(void)
 {
+#ifdef __REACTOS__
+    HWND hwndWorker;
+    EnterCriticalSection(&SHELL32_ChangenotifyCS);
+    hwndWorker = DoGetNewDeliveryWorker();
+    SendMessageW(hwndWorker, WM_NOTIF_REMOVEBYPID, GetCurrentProcessId(), 0);
+    LeaveCriticalSection(&SHELL32_ChangenotifyCS);
```

Comment:
What other code is this protecting against and what guarantees that said code won't try to access the critical section after it's been deleted?
(lock-unlock-delete is usually a sign that the synchronization design is not safe)

Reply:
I deleted some critical sections. I don't know sync well...